### PR TITLE
SNOW-1690717 Applying Snowpark Python function (sin)

### PR DESCRIPTION
**Pull request by sfc-gh-azhan**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   SNOW-1690717

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

This is the first PR to support Snowpark Python functions in pandas apply. It only introduce `sin` as the first example.

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 1]
```diff
@@ -1356,3 +1363,15 @@ def groupby_apply_sort_method(
             else GroupbyApplySortMethod.GROUP_KEY_APPEARANCE_ORDER
         )
     )
+
+
+def is_supported_snowpark_python_function(func: AggFuncType) -> bool:
+    """Return True if the `func` is a supported Snowpark Python function."""
+    func_module = inspect.getmodule(func)
+    if functions != func_module:
```
Comment (sfc-gh-jkew) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:1371]: Doesn't seem like checking the module is necessary if you are checking the function reference as well, but it probably doesn't hurt. Were you concerned about something in particular here?

**Review by sfc-gh-jkew — APPROVED** [review]

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 2]
```diff
@@ -8292,6 +8293,21 @@ def apply(
                 "Snowpark pandas apply API doesn't yet support DataFrame or Series in 'args' or 'kwargs' of 'func'"
             )
 
+        if is_supported_snowpark_python_function(func):
```
Comment (sfc-gh-joshi) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:8296]: do we need to add this in `map` as well?

File: tests/integ/modin/test_apply_snowpark_python_functions.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        s.to_frame().applymap(sin), native_s.to_frame().applymap(math.sin)
+    )
+    assert_frame_equal(
+        s.to_frame().apply(sin),
+        native_s.to_frame().apply(np.sin),  # Note math.sin does not work with df.apply
+    )
+
+
+@sql_count_checker(query_count=0)
+def test_apply_snowpark_python_function_not_implemented():
```
Comment (sfc-gh-joshi) [inline tests/integ/modin/test_apply_snowpark_python_functions.py:35]: should this test be parametrized instead?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 2 above)
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:8296]: `map` will call `applymap`

File: tests/integ/modin/test_apply_snowpark_python_functions.py (hunk 3 above)
Comment (sfc-gh-azhan) [reply tests/integ/modin/test_apply_snowpark_python_functions.py:35]: not necessary because it is super fast. No I/O or query execution.

**Review by sfc-gh-joshi — APPROVED** [review]

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py (hunk 1 above)
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:1371]: I just use this to report NotImplementedError for those not in the supported list.

**Pull request merged**